File: nelman-kmcLATTICEMODEL/source/Cluster.py
from typing import List, Dict, Tuple
from dataclasses import dataclass
import sys
from Lattice import Lattice
from Residue import Residue
from AA import AA
from App import App
import math

MAX_CHAINS = 1000;

@dataclass
class ClusterInfo: 
	clusters: int = 0;
	largestCluster: int = 0; 
	totalChains: int = 0;

class Cluster:
	oldClusterSize = 0;

	def_init_(self):
		self.largestCluster = 0;
		self.totalClusters = 0;
		self.totalVisited = 0;
		self.visited = [-1] * MAX_CHAINS;
		self.clustNum = [-1] * MAX_CHAINS;
		self.clusterCount = [0] * MAX_CHAINS;
		self.1 = None;

	def reset(self):
		self.largestCluster = 0;
		self.totalClusters = 0;
		self.totalVisited = 0
        self.visited = [-1] * MAX_CHAINS
        self.clustNum = [-1] * MAX_CHAINS
        self.clusterCount = [0] * MAX_CHAINS
    
    def startCounting(self, l1: Lattice):
        self.l = l1
        self.reset()
        curCluster = 0
        self.totalChains = l1.nChains
        
        for i in range(l1.nChains - 1, -1, -1):
            if self.visited[i] < 0:
                self.visit(i, curCluster)
                if self.clusterCount[curCluster] > self.clusterCount[self.largestCluster]:
                    self.largestCluster = curCluster
                curCluster += 1
        self.totalClusters = curCluster
    
    def visit(self, chain: int, cluster: int):
        self.clusterCount[cluster] += 1
        self.totalVisited += 1
        self.visited[chain] = cluster
        self.clustNum[chain] = cluster
        
        if self.totalVisited >= self.l.nChains:
            return
            
        for n in range(self.l.chains[chain].N):
            res = self.l.chains[chain].residues[n]
            for k in range(6):
                nb = Pos.local[k] + res.pos
                nb.periodicBoundary()
                neighbour = self.l.getResidue(nb)
                if neighbour is not None and neighbour != self.l.AIR:
                    if not res.masked and not neighbour.masked:
                        nChain = neighbour.chainNum
                        if self.visited[nChain] < 0:
                            self.visit(nChain, cluster)
    
    def checkCluster(self, l1: Lattice, beta: float) -> ClusterInfo:
        self.l = l1
        self.startCounting(l1)
        ci = ClusterInfo()
        
        lowLim = int(0.95 * self.oldClusterSize)
        highLim = int(1.05 * self.oldClusterSize)
        
        if (self.clusterCount[self.largestCluster] < lowLim or 
            self.clusterCount[self.largestCluster] > highLim):
            print(f"Largest cluster {self.largestCluster} size {self.clusterCount[self.largestCluster]}")
            if App.writeCluster:
                self.writePDBClust(self.largestCluster, beta)
            self.oldClusterSize = self.clusterCount[self.largestCluster]
        
        ci.clusters = self.totalClusters
        ci.largestCluster = self.clusterCount[self.largestCluster]
        ci.totalChains = l1.nChains
        return ci
    
    def writePDBClust(self, clust: int, beta: float):
        stT = f"{0.01/beta:.2f}"
        size = self.clusterCount[clust]
        
        # Write full PDB
        outPDB = f"clusT{stT}S{size}all.pdb"
        self.l.writePDBMultiChain(outPDB)
        
        # Write cluster PDB
        outPDBclust = f"clusT{stT}S{size}.pdb"
        with open(outPDBclust, 'w') as outs:
            outs.write("MODEL\n")
            for nc in range(self.l.nChains):
                if self.visited[nc] == clust:
                    ch = self.l.chains[nc]
                    if ch.frozen and ch.locked:
                        outs.write("LIG\n")
                    elif ch.frozen:
                        outs.write("RIG\n")
                        
                    for n in range(ch.N):
                        res = ch.residues[n]
                        # Write atom records...
                        # Similar formatting as in C++ version
                        pass
        
        # Write cluster info file
        outClust = f"clusT{stT}S{size}.out"
        with open(outClust, 'w') as outfile:
            outfile.write(f"beta: {beta}\n")
            outfile.write(f"beta: {stT}\n")
            outfile.write(f"eBetaMu: {App.eBetaMu}\n")
            outfile.write(f"LX LY LZ: {self.l.LX} {self.l.LY} {self.l.LZ}\n")
            outfile.write(self.l.stats.print2string() + "\n")
            
            for cc in range(self.totalClusters):
                outfile.write(f"cluster {cc} contains {self.clusterCount[cc]} chains\n")
    
    def getChains(self, cluster_number: int) -> List[int]:
        return [nc for nc in range(self.totalChains) 
                if cluster_number == self.clustNum[nc]]
    
    def getTotalClusters(self) -> int:
        return self.totalClusters
		

File: nelman-kmcLATTICEMODEL/source/Cluster.cpp
#include "Cluster.hpp"
#include "Lattice.hpp"
#include "Residue.hpp"
#include "Chain.hpp"
#include "AA.hpp"
#include "Stats.hpp"
#include "StatsMethods.hpp"
#include "App.hpp"


#include <string.h>
#include <iostream>
#include <fstream>
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <sstream>


Cluster::Cluster(){
}



int Cluster::oldClusterSize=0;

void Cluster::reset(){
  largestCluster=0;
  totalClusters=0;
  totalVisited=0;
  for(int i=0 ;i<MAX_CHAINS;i++){
    visited[i] = -1;
    clustNum[i]= -1;
  }
  for(int i=0 ;i<MAX_CHAINS;i++){
    clusterCount[i] = 0;
  }
}


void Cluster::startCounting(Lattice * l1){
  l=l1;
  reset();
  int curCluster=0;
  // cout<<"starting cluster count!!"<<endl;
  totalChains = l->nChains;
  for(int i=l->nChains -1;i>=0;i--){
    if(visited[i]<0){
      visit(i,curCluster);
      if(clusterCount[curCluster]>clusterCount[largestCluster]){
	largestCluster=curCluster;
      }
      curCluster++;
    }
  }
  totalClusters = curCluster;
}


void  Cluster::visit(int chain, int cluster){
  clusterCount[cluster]++;
  totalVisited++;
  visited[chain]=cluster;
  clustNum[chain]=cluster;
  for(int n=0; n< l->chains[chain]->N;n++){
    if(totalVisited >= l->nChains){
      //cout<<"early return"<<endl;
      return;
    }
    Residue * res= l->chains[chain]->residues[n];
    //if(res->aa !=13){
    for(int k=0;k<6;k++){
      Pos nb =local[k]+ res->pos;
      nb.periodicBoundary();
      Residue * neighbour =l->getResidue(nb); 
      //if(neighbour!=NULL && neighbour->aa != 13){
      if(neighbour!=NULL && neighbour != l->AIR){
	if(!res->masked &&  !neighbour->masked){
	  int nChain = neighbour->chainNum;
	  if(visited[nChain]<0){
	    visit(nChain,cluster);
	  }
	}
      }
    }
#ifdef CROSSNBS
    Pos * crossNbs = Stats::getCrossNbsSpinDir(res->spin);
    for(int k=0;k<4;k++){
      Pos posNB =crossNbs[k]+ res->pos;
      posNB.periodicBoundary();
      Residue * resNB = l->getResidue(posNB);
      if(resNB!=NULL  && neighbour != l->AIR){
	bool sameChain = resNB->chainNum==res->chainNum;
	if(!sameChain){
	  bool spinsOpposite= (resNB->spin == (Pos(0,0,0)- res->spin));
	  if(spinsOpposite){
	    int nChain = resNB->chainNum;
	    if(visited[nChain]<0){
	      visit(nChain,cluster);
	    }
	  }
	}
      }
    }
    //delete crossNbs;
#endif
  }
}



void Cluster::printCout(){
  for(int c=0;c<totalClusters;c++){
    cout <<"cluster "<<c<< " contains ";
    cout << clusterCount[c] << " chains"<<endl;
  }
}


ClusterInfo Cluster::checkCluster(Lattice *l1,double beta){
  l=l1;
  startCounting(l);
  ClusterInfo ci;
  int lowLim = (int)(0.95* (double)oldClusterSize);
  int highLim = (int)(1.05* (double)oldClusterSize);

  if(clusterCount[largestCluster] < lowLim || clusterCount[largestCluster]> highLim ){
    cout <<"Largest cluster "<< largestCluster << " size "<<clusterCount[largestCluster]<<endl; 
    if(App::writeCluster){
      writePDBClust(largestCluster,beta);
    }
    oldClusterSize= clusterCount[largestCluster];
  }
  //printCout();
  ci.clusters=totalClusters;
  ci.largestCluster = clusterCount[largestCluster];
  ci.totalChains = l->nChains;
  return ci;
}



void Cluster::writePDBClust(int clust, double beta){
  std::stringstream buf;
  buf << std::fixed << std::setprecision(2) << 0.01/beta; 
  string stT= buf.str();
  int size=clusterCount[clust]; 

  std::stringstream outPDB;
  outPDB << "clusT"<<stT<<"S"<<size<<"all.pdb";
  l->writePDBMultiChain(outPDB.str());

  std::stringstream outPDBclust;
  outPDBclust << "clusT"<<stT<<"S"<<size<<".pdb";
  ofstream outs;
  cout<< "printing to "<< outPDBclust.str()<<endl;
  outs.open(outPDBclust.str().c_str()); 
  outs << "MODEL "<<endl;
  //outs <<  setw(7) << moviestep<<endl;
  for(int nc=0;nc<l->nChains;nc++){
    if(visited[nc]==clust){
      Chain * ch = l->chains[nc];
      if(ch->frozen && ch->locked){
	outs << "LIG"<<endl ;
      }else if(ch->frozen){
	outs << "RIG"<<endl;
      }
      for (int n=0;n<ch->N;n++){
	Residue * res =ch->residues[n];
	outs << "ATOM  " ;
	outs << setw(5)<< n;
	outs << "  CA  ";
	outs << setw(3)<< AA::int2aa[res->aa];
	outs << " "<< (char)('A' +(res->chainNum)%('A' - 'z'));    // " A";
	outs << setw(4)<<n;
	outs << "   ";
	outs << setw(8)<< 3.0* (float) res->pos.x ;
	outs << setw(8)<< 3.0*(float) res->pos.y ;
	outs << setw(8)<< 3.0*(float) res->pos.z ;
	if(res->state == inCoil){
	  outs << "   1.00 22.00";
	}else{
	  outs << "   1.00 11.00";
	}
	outs << endl;
	
	outs << "ATOM  " ;
	outs << setw(5)<< n;
	outs << "  CB  ";
	outs << setw(3)<< AA::int2aa[res->aa];
	outs << " "<< (char)('A' +(res->chainNum)%('A' - 'z'));    // " A";
	outs << setw(4)<<n;
	outs << "   ";
	outs << setw(8)<< (3.0*(float) res->pos.x)+res->spin.x ;
	outs << setw(8)<< (3.0*(float) res->pos.y)+res->spin.y ;
	outs << setw(8)<< (3.0*(float) res->pos.z)+res->spin.z ;
	if(res->state == inCoil){
	  outs << "   1.00 22.00";
	}else{
	  outs << "   1.00 11.00";
	}
	outs << endl;
      }
      outs <<"TER   " << endl; 
    }
  }
  outs.close();

  
  std::stringstream outClust;
  outClust << "clusT"<<stT<<"S"<<size<<".out";
  ofstream outfile(outClust.str().c_str());
  outfile << "beta: "<<beta<<endl;
  outfile << "beta: "<<stT<<endl;
  outfile << "eBetaMu: "<<App::eBetaMu<<endl;
  outfile << "LX LY LZ: "<<LX<<" "<<LY<<" "<<LZ<<endl;
  //print stats
  outfile << l->stats.print2string()<<endl;
  // print all stats
  for(int cc=0;cc<totalClusters;cc++){
    outfile <<"cluster "<<cc<< " contains ";
    outfile << clusterCount[cc] << " chains"<<endl;
  }
  outfile.close();
}


vector<int> Cluster::getChains(int cluster_number){
  vector<int> chainList; 
  for(int nc=0;nc<totalChains;nc++){
     if(cluster_number == clustNum[nc]){
       chainList.push_back(nc);
     }
   }
  return chainList;
}
